core/src/floodlight_modules/busylink_detect.py

```diff
@@ -4,181 +4,185 @@
+            return 10*(1024**2)/8.0            #10Mb
+        return 0
+
+    def parsePortFeatures_v1_3(self,features):
+        if features == 0:
+            return 0
+        turn_binary = bin(features)[2:]
+        binary_len = len(turn_binary)
+        if binary_len < 12:
+            turn_binary = '0'*(12-binary_len) + turn_binary
+
+        if turn_binary[2] == '1':
+            return 1024**4/8.0                 #1Tb
+        if turn_binary[3] == '1':
+            return 100*(1024**3)/8.0           #100Gb
+        if turn_binary[4] == '1':
+            return 40*(1024**3)/8.0            #40Gb
+        if turn_binary[5] == '1':
+            return 10*(1024**3)/8.0            #10Gb
+        if turn_binary[6] == '1' or turn_binary[7] == '1':
+            return 1024**3/8.0                 #1Gb
+        if turn_binary[8] == '1' or turn_binary[9] == '1':
+            return 100*(1024**2)/8.0           #100Mb
+        if turn_binary[10] == '1' or turn_binary[11] == '1':
+            return 10*(1024**2)/8.0            #10Mb
+        return 0
+
+    def busyLinkDetect(self):
+        self.BLD_result = []
+        # Calculate link's countBytes and capacity
+        for link_id in self.links:
+            src = self.links[link_id]['source']
+            srcp = self.links[link_id]['sourcePort']
+            dest = self.links[link_id]['target']
+            destp = self.links[link_id]['targetPort']
+
+            # Check if needed information all arrived
+            if (src not in self.switches) | (dest not in self.switches):
+                print 'Not Ready'
+                return
+            elif (srcp not in self.switches[src]) | (destp not in self.switches[dest]):
+                print 'Not Ready'
+                return
+
+            total_bytes = self.switches[src][srcp] + self.switches[dest][destp]
+            self.links[link_id]['countBytes'] = total_bytes
+            self.links[link_id]['capacity'] = min(self.capacity["%s_%d" % (src,srcp)],self.capacity["%s_%d" % (dest,destp)])
+
+        # Initialize self.statistics value
+        if len(self.statistics) == 0:
+            self.statistics = dict(self.links)
+            for link_id in self.statistics:
+                self.statistics[link_id]['state'] = self.baseState
+
+        # Check threshold
+        for link_id in self.links:
+            if link_id in self.statistics:
+                if (self.links[link_id]['countBytes'] - self.statistics[link_id]['countBytes']) / self.statistics[link_id]['capacity'] >= self.threshold:
+                    self.overthreshold(link_id)
+                else:
+                    self.underthreshold(link_id)
+                self.statistics[link_id]['countBytes'] = self.links[link_id]['countBytes']
+            else:
+                self.statistics[link_id] = dict(self.links[link_id])
+                self.statistics[link_id]['state'] = self.baseState
+        # Remove unexisted link info
+        for link_id in self.statistics:
```

Comment:
Add/removing elements in a dictionary while iterating through it could lead to runtime error like the following:

~~~
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/vagrant/.local/lib/python2.7/site-packages/omniui-1.0.0-py2.7.egg/src/core.py", line 266, in iterate
    handler.handler(event)
  File "/home/vagrant/.local/lib/python2.7/site-packages/omniui-1.0.0-py2.7.egg/src/floodlight_modules/busylink_detect.py", line 79, in getPort
    self.busyLinkDetect()
  File "/home/vagrant/.local/lib/python2.7/site-packages/omniui-1.0.0-py2.7.egg/src/floodlight_modules/busylink_detect.py", line 173, in busyLinkDetect
    for link_id in self.statistics:
RuntimeError: dictionary changed size during iteration
~~~